### PR TITLE
ARROW-18087: [C++] RecordBatch::Equals should not ignore field names

### PR DESCRIPTION
**Pull request by jorisvandenbossche**

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
Failures seem to be unrelated

File: cpp/src/arrow/dataset/test_util.h [hunk 1]
```diff
@@ -157,7 +157,7 @@ class DatasetFixtureMixin : public ::testing::Test {
     std::shared_ptr<RecordBatch> lhs;
     ASSERT_OK(expected->ReadNext(&lhs));
     EXPECT_NE(lhs, nullptr);
-    AssertBatchesEqual(*lhs, batch, true);
+    AssertBatchesEqual(*lhs, batch);
```
Comment (pitrou) [inline cpp/src/arrow/dataset/test_util.h:160]: Why the change here? Behavior is probably not changed for `check_metadata=true`, only for `check_metadata=false`.

**Comment by pitrou** [comment]
@jorisvandenbossche Can you also rebase to get CI fixes from master?

File: cpp/src/arrow/dataset/test_util.h (hunk 1 above)
Comment (jorisvandenbossche) [reply cpp/src/arrow/dataset/test_util.h:160]: See https://github.com/apache/arrow/pull/14326#discussion_r998421185, this is only reverting what was added in that other PR (to avoid having that PR be blocked on this one). 
In theory, it shouldn't matter, because if it passes without true, it will certainly pass with `check_metadata=True`. It's just not needed in this case (so to simplify the code and avoid that people wonder why it is here)

Comment (pitrou) [reply cpp/src/arrow/dataset/test_util.h:160]: I see, thank you.

**Review by pitrou — APPROVED** [review]
